### DISCUSSION TITLE
Is it already foreseeable when MAUI can be used professionally to develop apps?

### DISCUSSION THREAD

**Discussion by DominikErnst** [dotnet/maui#21277] [Q&A] [answered]
### Description

I want a working MAUI development environment in which I can develop software professionally.

- I need a working debugging (Why do I see binding errors but the app works?)
- No pro table control
- Why is switching from one tab to another so slow? 
- Why are there pointless error messages "Binding could be compiled if x:DataType is specified." It is specified?!?!
- Why does Chat GPT always give me answers for Xamarin?
- ... and 1000 other frustrating problems!

This is like software development in the 90s!

### Steps to Reproduce

Use MAUI and be frustrated!

### Version with bug

8.0.7 SR2

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

8.0.7 SR2

### Affected platforms

Windows

### Did you find any workaround?

NO

**Comment by patrickklaeren** [comment]
I don't think this is particularly productive as an issue, let alone a discussion.

> I need a working debugging (Why do I see binding errors but the app works?)

I think this is circumstantial. Likewise in WPF, and co, Binding errors/warnings can exist because of ordering of initialisation, race conditions or merely to let you know something is non optimal, i.e. conversions from null etc. Historically these have always been "maybe there's a problem". Depends if they're internal or not.

> No pro table control

Not really sure what to do with this one. There are third party component partners that have tables, as well as being able to build one in Blazor and use in Hybrid? I suspect you're after a better data grid/table solution? They're usually never a good answer on mobile layouts.

> Why is switching from one tab to another so slow?

I think again this is circumstantial. Maybe share a demo and others could look at it?

> Why are there pointless error messages "Binding could be compiled if x:DataType is specified." It is specified?!?!

Not sure I've personally seen pointless error messages, this also looks like a warning to me.

> Why does Chat GPT always give me answers for Xamarin?

Depending on what you're using, you're using a model that was trained in a time where Xamarin was the goto. MAUI is still relatively new. Not much the MAUI team can do here.

> ... and 1000 other frustrating problems!

Quite vague. File issues and help the product forward.

**Reply by DominikErnst** [reply]
Debugging is bad.

I just need a simple table control that makes the columns the same width for each entry. Some comfort would be an automatic adjustment of the column width, but it should then be the same width for all entries.

?

x:DataType, I found out that it wants the data type of the BindingContext. I wouldn't have expected that at this point!

ChatGPT see below.

MS must invest more money to improve MAUI!

Thank you for your answers.

**Reply by ewerspej** [reply]
> I just need a simple table control that makes the columns the same width for each entry. Some comfort would be an automatic adjustment of the column width, but it should then be the same width for all entries.

Not all built-in controls can satisfy everyone's needs. There are plenty of third-party controls for this purpose. Checkout [Syncfusion](https://www.syncfusion.com/maui-controls), [DevExpress](https://www.devexpress.com/maui/) and others. They provide powerful things like DataGrids and other types of table views.

> x:DataType, I found out that it wants the data type of the BindingContext. I wouldn't have expected that at this point!

You shouldn't make that kind of assumption. The purpose of the `x:DataType` attribute is clearly described in the [compiled bindings](https://learn.microsoft.com/dotnet/maui/fundamentals/data-binding/compiled-bindings) documentation.

MAUI may still have issues, but you cannot blame every issue you face solely on bugs or insufficiencies. I've been working with MAUI for almost two years now and while I've encountered and reported many problems, I've also found workarounds and alternative solutions to most of them.

**Reply by DominikErnst** [reply]
It's the year 2024! 20 years ago MAUI would have been good, but today I expect better from MS than an alpha version. Many companies today are dependent on software development. It simply has to work at a professional level. Releasing junk like this drives developers to the competition. 

I don't like controls from third-party manufacturers, they often have a different mindset to MS. I tried Telerik for a Blazor Project. What can I say, garbage! It took me more time to figure out how to work around restrictions than it took me to implement my business logic. The Telerik examples are great and work fine, but in a **_real app_** it was a disaster. 

> I've also found workarounds and alternative solutions to most of them.
-> You shouldn't have to!

Ok, thanks, I'll read through the documentation! Sometimes it does help ;-)

**Reply by ewerspej** [reply, marked as answer]
I work with it professionally.

Yes, it involves a lot of frustration, but it's been improving a lot over the past year. The most annoying issues by now involve the tooling and not the quality of MAUI.

I still think that MAUI has a great concept and I love the new possibilities it brings.

I'm currently migrating some XF apps to MAUI and they work much smoother and faster. I can even use less third-party controls now. And every professional app in one way or another relies on workarounds, there's no perfect framework.

It's in flux, there are major improvements and the community is very actively supporting the dev team in various ways. I want MAUI to succeed so I participate instead of only criticizing all the time.

**Reply by DominikErnst** [reply]
You're absolutely right, but I need a mature development environment now, not next year.

**Reply by DominikErnst** [reply]
I found this from on Jun 29, 2022 https://github.com/dotnet/maui/issues/8433. It's still not solved. The last entry is from Nov 24, 2023.

_**And this is supposed to be a professional development product?**_

**Comment by davidortinau** [comment]
@DominikErnst for errors and warnings as you've described please file VS feedback issues.

To improve your experience with GPT, add more context to your prompt to help guide it. That will improve with time, and as Copilot itself is improved (if you're using Copilot).

**Reply by ewerspej** [reply]
Exactly. Plus, it's necessary to use GPT 4, because 3.5 is still stuck in time.

**Reply by DominikErnst** [reply]
I use different Chats in ChatGPT, e.g. "Take on the role of a MAUI developer." In the "Custom Instructions" settings it also says I am a MAUI developer. Copilot is not nearly as good as ChatGPT 3.5!
GPT4 is too expensive! $5 would be ok, but $20?!?! I have to pay for it myself. :-(

**Comment by georg-jung** [comment]
Seems like this was moved from #21272 but somehow ended up in the Announcements category, where it probably doesn't belong (seems quite exposed there, which is why I ended up here in the first place; this showed up on my github.com homepage as an announcement). /cc @PureWeen as you originally converted the issue to a discussion if I get the GitHub UI right...

**Reply by PureWeen** [reply]
😢 whoops

Well that was an unfortunate mistake

Thank you for calling it out @georg-jung

**Comment by AmrAlSayed0** [comment]
> Why is switching from one tab to another so slow?

Have you tried increasing the [TabbedPage.OffscreenPageLimit](https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.controls.platformconfiguration.androidspecific.tabbedpage.offscreenpagelimit)?

**Reply by DominikErnst** [reply]
Microsoft.Maui.Controls.PlatformConfiguration.Android, so ... no!

**Comment by MAUIoxo** [comment]
@davidortinau - the original question was about when MAUI can be used professionally. Well, I am not saying it is not possible to use it professionally. But, what really is a pain sometimes is that there are still a lot of Bugs in layouts, `CollectionView` and other standard controls that are the bread and butter for every MAUI developer. Often enough combined with regression bugs and some nasty to find workarounds. I was trying to figure out the right scenarios from my App, which very often was not so easy, and repprted issues here. Well, we are doing it to help taking this project to a good state.

But, my problem really is, that after writing these issues nobody seems to care about them anymore for very long time. I still have unsolved issues from September and others have some of them even longer.

When I look at the fixes and checkins on this project on a daily basis, my impression is that the majority of checkins appear minor and far away from the bugs that are reported daily. To me this is a big problem regrading the subject „can we work professionally“. The basic controls and mechanisms should be clean so that we can work professionally. And if issues come up it should be possible to get it fixed in an acceptable time frame, otherwise people that doing it professionally are stuck. And if you have to pay people it is even worse and the money can kill everything.

I was lucky that I could find workarounds for things that did not work as intended. But it took me hours to days. And I could have invested this time better to get my App done 😉

I know this sounds like a big complaint. Yes a bit, but it is not meant in a nasty way. I just wanted to ask if it wasn’t possible to hire more people for this project to fix Bugs sooner or at all? 

I wrote Bugs beginning of September and now it is April. Working professionally from a developer perspective is the other way around. How would a company finish their products when they are waiting so long for stuff they can’t find a workaround for?

Then I also do not get it why a solution is so close and someone already provided a Pull request but it doesn’t find it’s way into the code and why there is no solution available 

https://github.com/dotnet/maui/pull/20346


My biggest wish for this project really is that the open Issues are scanned for things that relate to "Standard Controls", Layouts or Controls like `CollectionView`, `Grid` and other things that do not layout properly that bother on a daily basis and bringt them to good shape really fixing the issues. Then also look for things that are brought in that everyone passes the tests and also has a new test in the Checkin. There are so many things that are regression when reading over the Bugs. This keeps from getting new nasty bugs that consume time reading over it evaluating etc. and keeping from realizing the new features you have in mind. Sorry for all that mentioned 😉
